tests/containers/multi_runtime.pm

```diff
@@ -0,0 +1,118 @@
+# SUSE's openQA tests
+#
+# Copyright 2024-2025 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Validate docker and podman network interoperability
+# - validate if podman and docker networking works together
+# - check if firewalld doesn't break either
+# Maintainer: QE-C team <qa-c@suse.de>
+
+use Mojo::Base qw(containers::basetest);
+use testapi;
+use serial_terminal qw(select_serial_terminal select_user_serial_terminal);
+use utils;
+use containers::common qw(install_packages);
+use Utils::Logging 'save_and_upload_log';
+
+my @containers;
+my $test_image = "registry.opensuse.org/opensuse/nginx";
+
+sub run_tests {
+    my $test_url = "http://www.opensuse.org";
+    my $volumes = '-v $HOME/nginx/nginx.conf:/etc/nginx/nginx.conf:ro,z -v $HOME/nginx:/usr/share/nginx/html:ro,z';
+    my $curl_opts = "-svL -o /dev/null";
+    my $port = 8000;
+
+    # Create rootfull & rootless containers on both podman & docker
+    for my $runtime ("podman", "docker") {
+        for my $type ("root", "rootless") {
+            my $name = "test_${type}_${runtime}";
+            my $sudo = ($type eq "root") ? "sudo" : "";
+            script_retry("$sudo $runtime pull $test_image", retry => 3, delay => 60, timeout => 180);
+            assert_script_run "$sudo $runtime run -d --name $name -p $port:80 $volumes $test_image";
+            push @containers, {name => $name, runtime => $runtime, sudo => $sudo, port => $port};
+            $port++;
+        }
+    }
+
+    # Test connectivity to all containers on localhost
+    for my $container (@containers) {
+        assert_script_run "curl -4 $curl_opts http://127.0.0.1:$container->{port}",
```

Comment:
Why not move this into the for loop above? Then we can always create the container, run the tests and delete the container afterwards and the whole cleanup routine becomes obsolete. Also we then don't need to count the port up but can just use one.

Reply:
I plan to extend this to IPv6.  And prefer to run the basic tests first, then IPv4, then IPv6, and so on.

Reply:
The order can still be done in such a way, perhaps even a own subroutine?

My suggestion would make the cleanup routine unnecessary, and I think that should be considered.

Reply:
I refactored to cleanup at the end of `run_tests` and added a generic cleanup routine in the post fail hook that I need to collect logs and the ouput of `$runtime inspect`